content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+
+## 2. Benefit: overrides
+
+As a consequence from the parametrized builds, we can also change the value of
+the parameters after the fact, using the derivations `override` function.
+
+Consider this new `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  hello = pkgs.callPackage ./hello.nix { };
+  people = pkgs.callPackage ./hello.nix { audience = "people"; };
+  folks = hello.override { audience = "folks"; };
+}
+```
+
+Building and running the `folks` attribute, will give again a new version of the
+script.
+
+All the other parameters will remain the same as the have been when `hello` was
+instantiated.
+
+This is especially useful and often seen on packages that provide a whole lot of
+options to optimize the build.
+
+An example to mention here is the `neovim` attribute in nixpkgs, which has has 
+some overrideable arguments like `extraLuaPackages`, `extraPythonPackages`, or
+`withRuby`.
+
+## 3. Benefit: modifiable
+
+And now I want to introduce one of my most favorite benefits:
+
+You can actually create your own version of `callPackage`, which comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+```nix
+# In the next examples I will not implement or show the "called" 
+# files, as I think they are not necessary to understand the point I
+# want to make.
+```
+
+Consider the following initial version:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  a = pkgs.callPackage ./a.nix { };
+  b = pkgs.callPackage ./b.nix { inherit a; };
+  c = pkgs.callPackage ./c.nix { inherit b; };
+  d = pkgs.callPackage ./d.nix { };
+  e = pkgs.callPackage ./e.nix { inherit c d; };
+}
+```
+
+Here you have to remember passing required arguments that are not in nixpkgs'
+toplevel manually.
+
+This can become quite tedious quickly, especially the larger the set becomes.
+
+Therefore we can use `lib.callPackageWith` to create our own `callPackage` version.
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+  callPackage = lib.callPackageWith (pkgs // packages);
+  packages = {
+    a = callPackage ./a.nix { };
+    b = callPackage ./b.nix { };
+    c = callPackage ./c.nix { };
+    d = callPackage ./d.nix { };
+    e = callPackage ./e.nix { };
+  };
+in
+  packages
+```
+
+Our modified `callPackage` now will exactly "know" how to resolve the dependencies
+through the set defined by `pkgs // packages`.
+
+Nix' laziness does us a good favour here and makes this actually possible.
+
+## Summary
+
+So using `callPackage` doesn't only make your code uniform to what you see a lot
+in nixpkgs already, it also gives you some things for free:
+
+1. parametrized builds
+2. overrideable builds
+3. cleaner implementation of large interdepending package sets
+
+## Further reading
+
+There is also `callPackages` and `lib.callPackages` which do a pretty similar
+thing, though they expect that the return value is *not* a package, but a
+packageset.
```

Comment:
There is a fundamental source of confusion around derivations and packages, and I (barely) tried to be a bit more rigorous about it in my suggestions. I think using them interchangeably is not just plain wrong - because the Nix language has no notion of packages, `nixpkgs` seems to have a very weak definition of a package, at least no authoritative one I'm aware of (one can of course make it up by saying "a package is a file declaring a function where arguments constitute all dependencies of the resulting derivation"), and you explicitly abstain from flakes, which have a definition of package which I find of dubious merit - but also extremely confusing, even to experienced Nixers.

So in general I would strongly urge you to check to make that distinction clear and define the terms you use before you use them. And here specifically please do not introduce another term without defining it. I think one can just say attrset of pkgs, but if package set is a thing in `nixpkgs` it may be worthwhile introducing it.

```suggestion
There is also `callPackages` and `lib.callPackages` which do a pretty similar
thing, though they expect that the return value is *not* a package, but an attribute set of packages.
```

Reply:
Re: "derivation" vs. "package"… I am aware that the concepts are used quite interchangeably, though a big problem of mine is this:

As far as I understand, the "derivation" is actually the python-like `*.drv` file, which when "realized" becomes the built artifact, with a different hash in the store.

Though in the context of this blogpost, I actually wanted to use "derivation" for the result of the "builder" and "package" for the result of `callPackage`, I see that I am far away from this intend.

I will revise this later and in a seperate commit from the suggested smaller fixes.